File: TempoCore/Sources/TempoCore/CardTypes.swift
// 템포루틴 — 3카드 값 타입 (MASTER §5.5)
// @Model(앱 타깃)이 참조하는 순수 Codable 값 — Core→SwiftData 의존 금지(§5.10), 참조 방향은 앱→Core만.

import Foundation

// ① 일정 카드 — 절대 날짜 / 연 반복 (cycle-anchored 모델 밖, resolveDate 안 씀)
public enum ScheduleRepeat: String, Codable, Sendable, CaseIterable { case none, daily, weekly, monthly, yearly }

// ② Input 카드
public enum InputCategory: String, Codable, CaseIterable, Sendable { case food, exercise, media, other }

public enum InputSchedule: Codable, Equatable, Sendable {
    case daily                          // 매일
    case weekly                         // 매주 — 생성일(createdAt)과 같은 요일(2026-07-22 확장)
    case monthly                        // 매달 — 생성일과 같은 일(day), 말일 클램프(2026-07-22 확장)
    case cycleAnchored(CycleRecurrence) // 주기 기준 (resolveDate 사용)
}

// 연관값 enum auto-synthesis는 실기기서 불안정(§5.5.1 실측) → discriminator 커스텀 Codable.
extension InputSchedule {
    enum CodingKeys: String, CodingKey { case type, recurrence }
    enum Kind: String, Codable { case daily, weekly, monthly, cycleAnchored }
    public func encode(to e: Encoder) throws {
        var c = e.container(keyedBy: CodingKeys.self)
        switch self {
        case .daily:
            try c.encode(Kind.daily, forKey: .type)
        case .weekly:
            try c.encode(Kind.weekly, forKey: .type)
        case .monthly:
            try c.encode(Kind.monthly, forKey: .type)
        case .cycleAnchored(let r):
            try c.encode(Kind.cycleAnchored, forKey: .type)
            try c.encode(r, forKey: .recurrence)
        }
    }
    public init(from d: Decoder) throws {
        let c = try d.container(keyedBy: CodingKeys.self)
        switch try c.decode(Kind.self, forKey: .type) {
        case .daily:          self = .daily
        case .weekly:         self = .weekly
        case .monthly:        self = .monthly
        case .cycleAnchored:  self = .cycleAnchored(try c.decode(CycleRecurrence.self, forKey: .recurrence))
        }
    }
}

// ③ Output 카드 — 진행도 종류 (§5.5.2: 진행은 아이템 수명 누적, 완료는 파생)
public enum OutputProgressKind: String, Codable, Sendable { case subtasks, sessions, percent }

/// Output 반복 — InputSchedule과 동형(2026-07-22 확장, 주기 데이터 없어도 daily/weekly/monthly는 동작해야 함).
public enum OutputSchedule: Codable, Equatable, Sendable {
    case daily                          // 매일
    case weekly                         // 매주 — 생성일(createdAt)과 같은 요일
    case monthly                        // 매달 — 생성일과 같은 일(day), 말일 클램프
    case cycleAnchored(CycleRecurrence) // 주기 기준 (resolveDate 사용)
}

extension OutputSchedule {
    enum CodingKeys: String, CodingKey { case type, recurrence }
    enum Kind: String, Codable { case daily, weekly, monthly, cycleAnchored }
    public func encode(to e: Encoder) throws {
        var c = e.container(keyedBy: CodingKeys.self)
        switch self {
        case .daily:
            try c.encode(Kind.daily, forKey: .type)
        case .weekly:
            try c.encode(Kind.weekly, forKey: .type)
        case .monthly:
            try c.encode(Kind.monthly, forKey: .type)
        case .cycleAnchored(let r):
            try c.encode(Kind.cycleAnchored, forKey: .type)
            try c.encode(r, forKey: .recurrence)
        }
    }
    public init(from d: Decoder) throws {
        let c = try d.container(keyedBy: CodingKeys.self)
        switch try c.decode(Kind.self, forKey: .type) {
        case .daily:          self = .daily
        case .weekly:         self = .weekly
        case .monthly:        self = .monthly
        case .cycleAnchored:  self = .cycleAnchored(try c.decode(CycleRecurrence.self, forKey: .recurrence))
        }
    }
}

File: android/tempocore/src/main/kotlin/app/temporoutine/core/CardTypes.kt
// 템포루틴 — 3카드 값 타입 (MASTER §5.5)
// iOS TempoCore/CardTypes.swift 1:1 이식. 연관값 enum은 discriminator(`type`) 커스텀 직렬화(§5.5.1).

package app.temporoutine.core

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.descriptors.element
import kotlinx.serialization.encoding.CompositeDecoder
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.encoding.decodeStructure
import kotlinx.serialization.encoding.encodeStructure

// ① 일정 카드 — 절대 날짜 / 연 반복 (cycle-anchored 모델 밖)
@Serializable
enum class ScheduleRepeat {
    @SerialName("none") NONE,
    @SerialName("daily") DAILY,
    @SerialName("weekly") WEEKLY,
    @SerialName("monthly") MONTHLY,
    @SerialName("yearly") YEARLY;

    val rawValue: String get() = name.lowercase()

    companion object {
        /** Swift `init?(rawValue:)` 대응 — 봉투·엔티티의 문자열 컬럼에서 되읽을 때. */
        fun fromRawValue(raw: String): ScheduleRepeat? = entries.firstOrNull { it.rawValue == raw }
    }
}

// ② Input 카드
@Serializable
enum class InputCategory {
    @SerialName("food") FOOD,
    @SerialName("exercise") EXERCISE,
    @SerialName("media") MEDIA,
    @SerialName("other") OTHER;

    val rawValue: String get() = name.lowercase()

    companion object {
        fun fromRawValue(raw: String): InputCategory? = entries.firstOrNull { it.rawValue == raw }
    }
}

/** Input 반복. ONCE = 반복 없음(단발 체크). */
@Serializable(with = InputScheduleSerializer::class)
sealed class InputSchedule {
    data object Once : InputSchedule()
    data object Daily : InputSchedule()
    data object Weekly : InputSchedule()      // 생성일과 같은 요일
    data object Monthly : InputSchedule()     // 생성일과 같은 일, 말일 클램프
    data class CycleAnchored(val recurrence: CycleRecurrence) : InputSchedule()
}

/** Output 반복 — InputSchedule과 동형. */
@Serializable(with = OutputScheduleSerializer::class)
sealed class OutputSchedule {
    data object Once : OutputSchedule()
    data object Daily : OutputSchedule()
    data object Weekly : OutputSchedule()
    data object Monthly : OutputSchedule()
    data class CycleAnchored(val recurrence: CycleRecurrence) : OutputSchedule()
}

// ③ Output 카드 — 진행도 종류 (§5.5.2)
@Serializable
enum class OutputProgressKind {
    @SerialName("checkOnly") CHECK_ONLY,
    @SerialName("subtasks") SUBTASKS,
    @SerialName("sessions") SESSIONS,
    @SerialName("percent") PERCENT,
    @SerialName("timer") TIMER,
    @SerialName("stopwatch") STOPWATCH;

    val rawValue: String
        get() = when (this) {
            CHECK_ONLY -> "checkOnly"
            SUBTASKS -> "subtasks"
            SESSIONS -> "sessions"
            PERCENT -> "percent"
            TIMER -> "timer"
            STOPWATCH -> "stopwatch"
        }

    companion object {
        fun fromRawValue(raw: String): OutputProgressKind? = entries.firstOrNull { it.rawValue == raw }
    }
}

/** `{ "type": "...", "recurrence": {...} }` — iOS 커스텀 Codable과 동일 키. */
private object ScheduleJson {
    fun descriptor(name: String): SerialDescriptor = buildClassSerialDescriptor(name) {
        element<String>("type")
        element<CycleRecurrence>("recurrence", isOptional = true)
    }

    fun encode(encoder: Encoder, descriptor: SerialDescriptor, kind: String, recurrence: CycleRecurrence?) {
        encoder.encodeStructure(descriptor) {
            encodeStringElement(descriptor, 0, kind)
            if (recurrence != null) encodeSerializableElement(descriptor, 1, CycleRecurrence.serializer(), recurrence)
        }
    }

    fun decode(decoder: Decoder, descriptor: SerialDescriptor): Pair<String, CycleRecurrence?> =
        decoder.decodeStructure(descriptor) {
            var kind: String? = null
            var recurrence: CycleRecurrence? = null
            while (true) {
                when (val index = decodeElementIndex(descriptor)) {
                    0 -> kind = decodeStringElement(descriptor, 0)
                    1 -> recurrence = decodeSerializableElement(descriptor, 1, CycleRecurrence.serializer())
                    CompositeDecoder.DECODE_DONE -> break
                    else -> error("unexpected index $index")
                }
            }
            (kind ?: throw SerializationException("type missing")) to recurrence
        }
}

object InputScheduleSerializer : KSerializer<InputSchedule> {
    override val descriptor: SerialDescriptor = ScheduleJson.descriptor("InputSchedule")

    override fun serialize(encoder: Encoder, value: InputSchedule) = when (value) {
        InputSchedule.Once -> ScheduleJson.encode(encoder, descriptor, "once", null)
        InputSchedule.Daily -> ScheduleJson.encode(encoder, descriptor, "daily", null)
        InputSchedule.Weekly -> ScheduleJson.encode(encoder, descriptor, "weekly", null)
        InputSchedule.Monthly -> ScheduleJson.encode(encoder, descriptor, "monthly", null)
        is InputSchedule.CycleAnchored -> ScheduleJson.encode(encoder, descriptor, "cycleAnchored", value.recurrence)
    }

    override fun deserialize(decoder: Decoder): InputSchedule {
        val (kind, recurrence) = ScheduleJson.decode(decoder, descriptor)
        return when (kind) {
            "once" -> InputSchedule.Once
            "daily" -> InputSchedule.Daily
            "weekly" -> InputSchedule.Weekly
            "monthly" -> InputSchedule.Monthly
            "cycleAnchored" -> InputSchedule.CycleAnchored(recurrence ?: throw SerializationException("recurrence missing"))
            else -> throw SerializationException("unknown InputSchedule type: $kind")
        }
    }
}

object OutputScheduleSerializer : KSerializer<OutputSchedule> {
    override val descriptor: SerialDescriptor = ScheduleJson.descriptor("OutputSchedule")

    override fun serialize(encoder: Encoder, value: OutputSchedule) = when (value) {
        OutputSchedule.Once -> ScheduleJson.encode(encoder, descriptor, "once", null)
        OutputSchedule.Daily -> ScheduleJson.encode(encoder, descriptor, "daily", null)
        OutputSchedule.Weekly -> ScheduleJson.encode(encoder, descriptor, "weekly", null)
        OutputSchedule.Monthly -> ScheduleJson.encode(encoder, descriptor, "monthly", null)
        is OutputSchedule.CycleAnchored -> ScheduleJson.encode(encoder, descriptor, "cycleAnchored", value.recurrence)
    }

    override fun deserialize(decoder: Decoder): OutputSchedule {
        val (kind, recurrence) = ScheduleJson.decode(decoder, descriptor)
        return when (kind) {
            "once" -> OutputSchedule.Once
            "daily" -> OutputSchedule.Daily
            "weekly" -> OutputSchedule.Weekly
            "monthly" -> OutputSchedule.Monthly
            "cycleAnchored" -> OutputSchedule.CycleAnchored(recurrence ?: throw SerializationException("recurrence missing"))
            else -> throw SerializationException("unknown OutputSchedule type: $kind")
        }
    }
}
